---
title: "Aligning sequences and creating phylogenetic tree"
---

```{r}
#install and load required packages
#Bios2cor -- used to import fasta (fa) file into R
install.packages("Bios2cor")
library(Bios2cor)
#seqinr -- used to export fasta files from R
library(seqinr)
#readxl -- used to import excel files to R
#decipher -- used to create multiple alignment of the sequences 
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("DECIPHER")
library(DECIPHER)
library(Biostrings)
# dada2 package to handle sequence data
#BiocManager::install("dada2")
library(dada2)

#phangorn -- used to create a phylogenetic three
#install.packages("phangorn")
library(phangorn)
```
followed the toturial from http://web.stanford.edu/class/bios221/MicrobiomeWorkflowII.html
My sequences are stored in an .fa file. 

```{r}
filepath = "C:/Users/madel/OneDrive - NTNU/Doktorgrad/Project 8 Phage therapy and AB in semi-cont/R-analysis/GitHubScripts"
# import sequences 
file_location = "C:/Users/madel/OneDrive - NTNU/Doktorgrad/Project 8 Phage therapy and AB in semi-cont/R-analysis/GitHubScripts/rawdata/2022_09_14_zotus_16SrDNA-andRNA_Phage-AB-Fc7-Jonsvatne.fa"
sequences = Bios2cor::import.fasta(file = file_location)
#total of 2594 sequences 
#ASVs to remove based on contaminated sequences, archea or chloroplast
keep = readRDS(file = paste0(filepath, "/rawdata/2022-09-15_taxa_keep.RDS"))
seq_keep = sequences[keep$ASV] #filters all sequences that matches the names of zOTU_keep, 5008 ASVs
#saved sequences as fasta file. 
seqinr::write.fasta(sequences = seq_keep, names = names(seq_keep), 
                    file.out =  paste0(filepath, "/rawdata/2022-09-15_zOTUsequences_to_keep.fa"))
```

```{r}
#import the fasta file into object called seqs
seqs <- dada2::getSequences(object =paste0(filepath, "/rawdata/2022-09-15_zOTUsequences_to_keep.fa"))
# tip labels are zOTU names (i.e. zOTU1, zOTU10, ect)
#align zOTU sequences using the  DECIPHER package.
alignment <- DECIPHER::AlignSeqs(Biostrings::DNAStringSet(seqs), anchor=NA,verbose=FALSE)
saveRDS(object = alignment, file = paste0(filepath, "/results/2022-09-15_sequence_alignment.RDS"))
```

The phangorn R package is then used to construct a phylogenetic tree. Here we first construct a neighbor-joining tree, and then fit a GTR+G+I (Generalized time-reversible with Gamma rate variation) maximum likelihood tree using the neighbor-joining tree as a starting point.
```{r}
phangAlign <- phangorn::phyDat(as(alignment, "matrix"), type="DNA")
dm <- phangorn::dist.ml(phangAlign)
saveRDS(object = dm, file = paste0(filepath, "/results/2022-09-15_pylogenetic_dm.RDS"))
treeNJ <- phangorn::NJ(dm) # Note, tip order != sequence order
saveRDS(object = treeNJ, file = paste0(filepath, "/results/2022-09-15_pylogenetic_treeNJ.RDS"))
fit = pml(treeNJ, data=phangAlign)
fitGTR <- update(fit, k=4, inv=0.2)
#started 11:05

fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
        rearrangement = "stochastic", control = pml.control(trace = 1))

saveRDS(fitGTR, file = "C:/Users/madel/OneDrive - NTNU/Doktorgrad/Project 8 Phage therapy and AB in semi-cont/R-analysis/GitHubScripts/rawdata/2022-09-16_zOTU_phytree.RDS")
```

